opm/autodiff/SimulatorFullyImplicitBlackoilEbos.hpp

```diff
@@ -613,8 +619,205 @@ class SimulatorFullyImplicitBlackoilEbos
+            }
+            
+            for (int n = 0; n < 3; n++){
+                wellRate += wr[n][w];
+            }
+            if (is_producer) {
+                if (pu.phase_used[Water] && pu.phase_used[Oil]) {
+                    if (wellRate > 0.00001){
+                        //Watercut
+                        wr[5][w] = wr[0][w]/wellRate;
+                    }
+                }
+                if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                    //Gos-OilRatio (GOR)
+                    wr[6][w] = wr[2][w]/wr[1][w];
+                }
+                if (pu.phase_used[Water] && pu.phase_used[Gas]) {
+                    //Water-Gas Ratio (WGR)       
+                    wr[7][w] = wr[0][w]/wr[2][w];
+                }
+            }
+        }
+        return wr;
+    }
+   
+   //Compute and return produced and injected totals from wells per phase
+   std::vector<double> wellTotals(std::vector<std::vector<double>> wr,
+                                  const Wells* wells,
+                                  const Opm::PhaseUsage& pu)
+   {
+        int nw = wells->number_of_wells;
+        std::vector<double> wtotals (9,0.0);        
+        for (int w = 0; w < nw; ++w) { 
+           const bool is_producer = wells->type[w] == PRODUCER;
+           const bool is_injector = wells->type[w] == INJECTOR;
+           if (is_producer) {
+               //Field Oil produced
+               wtotals[0] += wr[1][w];
+               //Field Water produced
+               wtotals[1] += wr[0][w];
+               //Field Gas produced
+               wtotals[2] += wr[2][w];
+           }
+           if (is_injector){
+               //Field Oil injected
+               wtotals[6] += wr[1][w];
+               //Field Water injected
+               wtotals[7] += wr[0][w];
+               //Field Gas injected
+               wtotals[8] += wr[2][w];
+           }
+        }
+        for (int w = 0; w < nw; ++w) {
+           const bool is_producer = wells->type[w] == PRODUCER;
+           if (is_producer) {
+               if (pu.phase_used[Water] && pu.phase_used[Oil]) {
+                   //Field Watercut
+                   wtotals[3] = wtotals[1]/(wtotals[0] + wtotals[1] + wtotals[2]);
+               }
+               if (pu.phase_used[Oil] && pu.phase_used[Gas]) {
+                   //Field GOR
+                   wtotals[4] = wtotals[2]/wtotals[0];
+               }
+               if (pu.phase_used[Water] && pu.phase_used[Gas]) {
+                   //Field WGR       
+                   wtotals[5] = wtotals[1]/wtotals[2];
+               }
+	   }
+        }
+        return wtotals;
+   }
+   
+   //Output well report to the .PRT file
+   void outputWellReport(std::vector<std::vector<double>> wr,
+             std::vector<double> wtotals,
+             const Wells* wells,
+             WellState& xw,
+             const UnitSystem& units)
+    {           
+        int nw = wells->number_of_wells;
```

Comment:
At this point we would need to add the communication to gather the well variables and totals. There is still one problem though: `wells` will only contain the local wells and all of them. Could we maybe move this code where we also do the output of the well curves to the eclipse files?  That would be in [`BlackoilOutputWriter::writeTimeStepSerial`](/OPM/opm-simulators/blob/bf0937fae1f0ae72139569e6419e41d8cdbe5888/opm/autodiff/SimulatorFullyImplicitBlackoilOutput.cpp#L319). There is already a well report created for opm-output anyway. Maybe we could reuse some code?